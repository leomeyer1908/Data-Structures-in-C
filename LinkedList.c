#include "LinkedList.h"



void initList(LinkedList* list) {
    list->head = NULL;
    list->tail = NULL;
}

void pushBackList(LinkedList* list, void* value) {
    DoublyNode* newNode = getNewDoublyNode();
    newNode->value = value;
    list->tail = newNode;
    if (list->head == NULL) {
        list->head = newNode;
    }
    else {
        DoublyNode* currentNode = list->head;
        while (currentNode->next != NULL) {
            currentNode = currentNode->next;
        }
        currentNode->next = newNode;
        newNode->prev = currentNode;
    }
}

void removeNodeFromList(LinkedList* list, DoublyNode* node) {
    if (node->prev != NULL) {
        node->prev->next = node->next;
    } else {
        list->head = node->next;
    }

    if (node->next != NULL) {
        node->next->prev = node->prev;
    } else {
        list->tail = node->prev;
    }

    free(node);

}

void removeElementList(LinkedList* list, void* value) {
    DoublyNode* currentNode = list->head;
    while (currentNode != NULL) {
        if (currentNode->value == value) {
            removeNodeFromList(list, currentNode);
            break;
        }
        currentNode = currentNode->next;
    }
}

void destroyList(LinkedList* list) {
    DoublyNode* currentNode = list->head;
    DoublyNode* nextNode;
    while (currentNode != NULL) {
        nextNode = currentNode->next;
        free(currentNode);
        currentNode = nextNode;
    }

    list->head = NULL;
    list->tail = NULL;
}

void copyList(LinkedList* srcList, LinkedList* dstList) {
	if (srcList->head != NULL) {
		DoublyNode* dstNode = getNewDoublyNode();
		dstList->head = dstNode;
		DoublyNode* srcNode = srcList->head;
		while (srcNode != NULL) {
			dstNode->value = srcNode->value;
			if (srcNode->next != NULL) {
				DoublyNode* newDstNode = getNewDoublyNode();
				dstNode->next = newDstNode;
				dstNode = newDstNode;
			}
			srcNode = srcNode->next;
		}
	}
	else {
		dstList->head = NULL;
	}
}

