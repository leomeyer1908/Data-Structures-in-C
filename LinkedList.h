#ifndef LINKED_LIST_H
#define LINKED_LIST_H

#include <stdlib.h>
#include "DoublyNode.h"

typedef struct LinkedList {
    DoublyNode* head;
    DoublyNode* tail;
} LinkedList;

void initList(LinkedList* list);

void pushBackList(LinkedList* list, void* value);

void removeElementList(LinkedList* list, void* value);

void destroyList(LinkedList* list);


#endif