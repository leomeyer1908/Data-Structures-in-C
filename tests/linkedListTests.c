#include <stdio.h>
#include <stdint.h>
#include "../LinkedList.h"

void printList(LinkedList* list) {
    DoublyNode* currentNode = list->head;
    while (currentNode != NULL) {
        printf("%zu", (size_t)currentNode->value); 
        currentNode = currentNode->next;
        if (currentNode != NULL) {
            printf(", ");
        }
    }
    printf("\n");
}

int main() {
    LinkedList myList;
    initList(&myList);

    printf("Test Case: Empty List\n");
    printList(&myList);
    printf("Size: %zu\n", myList.size); 

    // Verify the elements are in the correct order
    printf("\nTest Case: Push Back Elements (10, 20, 30, 40, 50)\n");

    pushBackList(&myList, (void*) (uintptr_t) 10);
    pushBackList(&myList, (void*) (uintptr_t) 20);
    pushBackList(&myList, (void*) (uintptr_t) 30);
    pushBackList(&myList, (void*) (uintptr_t) 40);
    pushBackList(&myList, (void*) (uintptr_t) 50);

    printList(&myList);
    printf("Size: %zu\n", myList.size); // Print the size

    // Test Case: Remove Element from the Middle
    printf("\nTest Case: Remove Element from the Middle \n");
    removeElementList(&myList, (void*) (uintptr_t) 30);
    printList(&myList);
    printf("Size: %zu\n", myList.size); // Print the size

    // Test Case: Remove Element from the Beginning
    printf("\nTest Case: Remove Element from the Beginning\n");
    removeElementList(&myList, (void*) (uintptr_t) 10);
    printList(&myList);
    printf("Size: %zu\n", myList.size); // Print the size

    // Test Case: Remove Element from the End
    printf("\nTest Case: Remove Element from the End\n");
    removeElementList(&myList, (void*) (uintptr_t) 50);
    printList(&myList);
    printf("Size: %zu\n", myList.size); // Print the size

    printf("\nTest Case: Copy List\n");
    LinkedList myList2;
    copyList(&myList, &myList2);

    printf("New List:\n");
    printList(&myList2); 
    printf("New List Size: %zu\n", myList2.size); // Print the size
    printf("Old List:\n");
    printList(&myList); 
    printf("Old List Size: %zu\n", myList.size); // Print the size

    printf("\nTest Case: Modifying old list does not affect new list\n");
    printf("Removing First 2 elements of old list then adding a new one:\n");
    removeElementList(&myList, (void*) (uintptr_t) 20);
    removeElementList(&myList, (void*) (uintptr_t) 40);
    pushBackList(&myList, (void*) (uintptr_t) 90);
    printf("New List:\n");
    printList(&myList2); 
    printf("New List Size: %zu\n", myList2.size); // Print the size
    printf("Old List:\n");
    printList(&myList); 
    printf("Old List Size: %zu\n", myList.size); // Print the size

    printf("\nTest Case: Modifying new list does not affect new list\n");
    printf("Destroying new list: \n");
    destroyList(&myList2);
    printf("New List:\n");
    printList(&myList2); 
    printf("New List Size: %zu\n", myList2.size); // Print the size
    printf("Old List:\n");
    printList(&myList); 
    printf("Old List Size: %zu\n", myList.size); // Print the size


    printf("\nTest Case: Adding duplicate elements then removing both\n");
    pushBackList(&myList, (void*) (uintptr_t) 10);
    pushBackList(&myList, (void*) (uintptr_t) 10);
    printList(&myList);
    printf("Size: %zu\n", myList.size); // Print the size
    removeElementList(&myList, (void*) (uintptr_t) 10);
    removeElementList(&myList, (void*) (uintptr_t) 10);
    printList(&myList);
    printf("Size: %zu\n", myList.size); // Print the size

    
    printf("\nTest Case: removing non existant element from list (10):\n");
    removeElementList(&myList, (void*) (uintptr_t) 100);
    printList(&myList);
    printf("Size: %zu\n", myList.size); // Print the size


    printf("\nTest Case: removing last element from list:\n");
    removeElementList(&myList, (void*) (uintptr_t) 90);
    printList(&myList);
    printf("Size: %zu\n", myList.size); // Print the size

    printf("\nTest Case: removing element from empty list:\n");
    removeElementList(&myList, (void*) (uintptr_t) 10);
    printList(&myList);
    printf("Size: %zu\n", myList.size); // Print the size

    printf("\nTest Case: Adding element and removing again one more time:\n");
    pushBackList(&myList, (void*) (uintptr_t) 10);
    printList(&myList);
    printf("Size: %zu\n", myList.size); // Print the size
    removeElementList(&myList, (void*) (uintptr_t) 10);
    printList(&myList);
    printf("Size: %zu\n", myList.size); // Print the size

    printf("\nTest Case: Destroying list, then adding element and removing again:\n");
    destroyList(&myList);
    printList(&myList);
    printf("Size: %zu\n", myList.size); // Print the size
    pushBackList(&myList, (void*) (uintptr_t) 10);
    printList(&myList);
    printf("Size: %zu\n", myList.size); // Print the size
    removeElementList(&myList, (void*) (uintptr_t) 10);
    printList(&myList);
    printf("Size: %zu\n", myList.size); // Print the size

    printf("\nTest Case: Destroying list again:\n");
    destroyList(&myList);
    printList(&myList);
    printf("Size: %zu\n", myList.size); // Print the size
    //No need to test removeNodeFromList function because it is already called in removeElementList

    return 0;
}