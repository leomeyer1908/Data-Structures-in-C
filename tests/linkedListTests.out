Test Case: Empty List

Size: 0

Test Case: Push Back Elements (10, 20, 30, 40, 50)
10, 20, 30, 40, 50
Size: 5

Test Case: Remove Element from the Middle 
10, 20, 40, 50
Size: 4

Test Case: Remove Element from the Beginning
20, 40, 50
Size: 3

Test Case: Remove Element from the End
20, 40
Size: 2

Test Case: Copy List
New List:
20, 40
New List Size: 2
Old List:
20, 40
Old List Size: 2

Test Case: Modifying old list does not affect new list
Removing First 2 elements of old list then adding a new one:
New List:
20, 40
New List Size: 2
Old List:
90
Old List Size: 1

Test Case: Modifying new list does not affect new list
Destroying new list: 
New List:

New List Size: 0
Old List:
90
Old List Size: 1

Test Case: Adding duplicate elements then removing both
90, 10, 10
Size: 3
90
Size: 1

Test Case: removing non existant element from list (10):
90
Size: 1

Test Case: removing last element from list:

Size: 0

Test Case: removing element from empty list:

Size: 0

Test Case: Adding element and removing again one more time:
10
Size: 1

Size: 0

Test Case: Destroying list, then adding element and removing again:

Size: 0
10
Size: 1

Size: 0

Test Case: Destroying list again:

Size: 0
